clang/test/SemaSYCL/array-num-elements-exceeded.cpp

```diff
@@ -0,0 +1,18 @@
+// RUN: %clang_cc1 -I %S/Inputs -fsycl -fsycl-is-device -verify -fsyntax-only %s
+#include "big_constant.h"
+
+template <typename Name, typename Func>
+__attribute__((sycl_kernel)) void kernel_single_task(Func kernelFunc) {
+
+  kernelFunc(); // expected-note {{called by 'kernel_single_task<fake_kernel, (lambda}}
+}
+
+int main() {
+  double hostResult = big[0]; // no error thrown, since accessing global const arrays with size > 65532 is valid from host code.
+  kernel_single_task<class fake_kernel>([]() {
+    double localArray[65588] = {0.0};
+    double kernelLocal = localArray[0]; // no error thrown, local variables inside kernel can have array size > 65532
```

Comment:
Why is it that locals are not affected by the limitation?

Reply:
This is because Global Variables have a minimum limit of `65,535`, while local variables have a minimum limit of `524,287`.
See [Table.3.](https://www.khronos.org/registry/spir-v/specs/1.0/SPIRV.html#_a_id_limits_a_universal_limits)
Also, accessing a global const array with size > 65532 from host code is fine, since there will be no SPIR-V generated for host code. So the limitation doesn't apply to host code.

Reply:
Yes, but the error you are generating is because not even the 65535 limit can be supported.  Wouldn't something with an even larger limit have the same problem?

Reply:
> but the error you are generating is because not even the 65535 limit can be supported.

Only global variables have this restriction of 65535 in SPIRV.
If you scroll down a bit further in that table, you can see that local variables have a higher limit. 
`Local variables (Function Storage Class) | 524,287`

The 65535 limit does not apply to local variables. 
Did I get your question right?

Reply:
But the limit in that tables are all minimum limits.  We are issuing diagnostics for global variables because we cannot support that limit.   Why is it not an issue for local variables?

Removing the check for local variables is not what I am asking.   I am asking if you should also be diagnosing it for local variables.

Reply:
I understand. Are you suggesting we create a local array exceeding the local variable limits and see if that emits an error?

Reply:
> But the limit in that tables are all minimum limits. We are issuing diagnostics for global variables because we cannot support that limit. Why is it not an issue for local variables?
> 
> Removing the check for local variables is not what I am asking. I am asking if you should also be diagnosing it for local variables.

If I understand correctly, the actual issue here is that SPIR-V instruction word count cannot exceed 65536. This caused an issue for global constant arrays with size > 65532 because in this case the corresponding SPIR-V instruction OpConstantComposite exceeds instruction word count limit. Maybe the instruction for local arrays is different and encodes array differently, and so doesn't cross this limit?

Reply:
exceed `65535`, but yes, that's my understanding. @AlexeySotkin @AlexeySachkov Can you confirm if this is correct?

Reply:
> Maybe the instruction for local arrays is different and encodes array differently, and so doesn't cross this limit?

Right, this is what I want confirmed.

Reply:
Do we make the difference between SPIR and SPIR-V?
Because this is a limitation of SPIR-V, not SPIR 1.2 or SPIR 2.0.
This matters because in our case we sometimes use SPIR 2.0 on FPGA.
But if you cannot easily makes the difference, I do not want to make your life too complex.
I guess it is unlikely we enter into this limit anyway and, if we do, we will make the change on our branch.

Reply:
`Context.getTargetInfo().getTriple().isSPIR()` only checks if the target is SPIR (32- or 64-bit).
I will have to check if we differentiate SPIR-V and SPIR.

Reply:
Note that the limit on local variables is the limit on the _number_ of local variables, while this diagnostic is about the _size_ of a constant array.

Reply:
>>I will have to check if we differentiate SPIR-V and SPIR.
I don't see a way to differentiate between SPIR and SPIRV targets. 
Will resolve this conversation. If there is a way, please let me know.